ftnchek.1
.lf 1 ftnchek.man
.TH "FTNCHEK 3.2" 1L "November 2002"
.\" ====================================================================
.\"  @Troff-man-file{
.\"     author          = "Robert K. Moniot",
.\"     version         = "3.2",
.\"     date            = "20 Nov 2002"
.\"     time            = "12:00:00 EST",
.\"     filename        = "ftnchek.man",
.\"     address         = "Fordham University
.\"                        New York, NY 10023
.\"                        USA",
.\"     telephone       = "+1 212 636 6311",
.\"     email           = "moniot@fordham.edu",
.\"     codetable       = "ISO/ASCII",
.\"     keywords        = "Fortran 77, syntax check, type declarations",
.\"     supported       = "yes",
.\"     docstring       = "This file is the source for the UNIX manual
.\"                        page document (ftnchek.1), and is also used to
.\"                        generate the flat ASCII version of the
.\"                        documentation (ftnchek.doc), the PostScript
.\"                        version (ftnchek.ps), the HTML version, the blurb
.\"                        used for informational mailings (blurb.txt), and
.\"                        the VMS HELP document (ftnchek.hlp) used to make a
.\"                        HELP library file.  For the latter, define the
.\"                        number register ``h'' to be nonzero (nroff -rh1),
.\"                        and filter the output thru col (see the Makefile).
.\"                        The blurb is created by setting number register
.\"                        ``b'' nonzero (nroff -rb1).  The different text
.\"                        blocks for these variations are set off using
.\"                        .if reg-name ... where reg-name is
.\"                        \nb for the blurb and \nh for the HELP document.
.\"                        The HTML version is produced using rman, whose
.\"                        output is lightly edited (using patch) and then
.\"                        split up and put into the html directory by
.\"                        the shell script make_html.sh.
.\"
.\"                        If the next line after these comments is not
.\"                        ``.ie \nh \{'' then the HELP and blurb
.\"                        materials have been removed and this file is
.\"                        ftnchek.1.  You should edit only ftnchek.man
.\"                        and remake ftnchek.1 et al from it."
.\"
.\"  }
.\" ====================================================================
.SH "NAME"
ftnchek \- Fortran 77 program checker
.SH "SYNOPSIS"
.B ftnchek
[
.BR \-arguments [=\fIlist\fP]
]
[
.BR \-array [=\fIlist\fP]
]
.if n .ti +5
[
.BR \- [ no ] brief
]
[
.BR \-calltree [=\fIlist\fP]
]
[
.BR \- [ no ] check
]
.if n .ti +5
.if t .ti +.5i
[
.BR \-columns [=\fInum\fP]
]
[
.BR \-common [=\fIlist\fP]
]
.if n .ti +5
[
.BR \- [ no ] crossref [=\fIlist\fP]
]
[
.BR \- [ no ] declare
]
.if n .ti +5
.if t .ti +.5i
[
.BR \- [ no ] division
]
[
.BR \-errors [=\fInum\fP]
]
[
.BR \- [ no ] extern
]
.if n .ti +5
[
.BR \- [ no ] f77 [ = \fIlist\fP]
]
[
.BR \- [ no ] f90 [ = \fIlist\fP]
]
.if n .ti +5
.if t .ti +.5i
[
.BR \- [ no ] f95 [ = \fIlist\fP]
]
[
.BR \- [ no ] help
]
.if n .ti +5
[
.BR \- [ no ] identifier\-chars [ = \fIlist\fP]
]
[
.BI \-include= \fIstr\fP
]
.if n .ti +5
.if t .ti +.5i
[
.BR \-intrinsic [=\fIlist\fP]
]
[
.BR \- [ no ] library
]
[
.BR \- [ no ] list
]
.if n .ti +5
[
.BR \-makedcls [=\fIlist\fP]
]
[
.BR \-mkhtml [=\fIlist\fP]
]
.if t .ti +.5i
.if n .ti +5
[
.BR \- [ no ] novice
]
[
.BI \-output= str
]
.if n .ti +5
[
.BR \-pointersize [=\fInum\fP]
]
[
.BR \- [ no ] portability [ = \fIlist\fP]
]
.if t .ti +.5i
.if n .ti +5
[
.BR \- [ no ] pretty [ = \fIlist\fP]
]
[
.BR \-project [=\fIlist\fP]
]
.if n .ti +5
[
.BR \- [ no ] pure
]
[
.BR \- [ no ] quiet
]
[
.BR \- [ no ] reference
]
.if t .ti +.5i
.if n .ti +5
[
.BR \- [ no ] resources
]
[
.BR \- [ no ] sixchar
]
[
.BR \- [ no ] sort
]
.if n .ti +5
[
.BR \-source [=\fIlist\fP]
]
[
.BR \-style [=\fIlist\fP]
]
.if t .ti +.5i
[
.BR \- [ no ] symtab
]
.if n .ti +5
[
.BR \- [ no ] truncation [ = \fIlist\fP]
]
[
.BR \-usage [=\fIlist\fP]
]
.if n .ti +5
[
.BR \- [ no ] vcg
]
[
.BR \- [ no ] version
]
.if t .ti +.5i
[
.BR \- [ no ] volatile
]
.if n .ti +5
[
.BR \-wordsize [=\fInum\fP]
]
[
.BR \-wrap [=\fInum\fP]
]
[
.I files ...
]
.SH "DESCRIPTION"
.PP
\fBftnchek\fP (short for Fortran checker)
is designed to detect certain errors in a
Fortran program that a compiler usually does not.
\fBftnchek\fP is not primarily
intended to detect syntax errors.  Its purpose is to assist the user in finding
semantic errors.  Semantic errors are legal in the Fortran language but are
wasteful or may cause incorrect operation.  For example, variables which are
never used may indicate some omission in the program; uninitialized variables
contain garbage which may cause incorrect results to be calculated; and
variables which are not declared may not have the intended type.
\fBftnchek\fP is
intended to assist users in the debugging of their Fortran program.  It is not
intended to catch all syntax errors.  This is the function of the compiler.
Prior to using \fBftnchek\fP,
the user should verify that the program compiles
correctly.
.\" Omit the following paragraph from the blurb
.PP
This document first summarizes how to invoke \fBftnchek\fP.
That section should be read before beginning to use \fBftnchek\fP.
Later sections describe \fBftnchek\fP's
options in more detail, give an example of its use, and explain
how to interpret the output.  The final sections mention the limitations
and known bugs in
\fBftnchek\fP.
.\" End of paragraph omitted from the blurb
.SH "INVOKING FTNCHEK"
\fBftnchek\fP is
invoked through a command of the form:
.sp 1
.ce 1
\fC$ ftnchek [-option -option ...] filename [filename ...]\fR
.PP
The brackets indicate something which is optional.  The brackets
themselves are not actually typed.  Here options are command-line
switches or settings, which control the operation of the program and the
amount of information that will be printed out.  If no option is
specified, the default action is to print error messages, warnings,
and informational messages, but not the program listing or symbol tables.
.PP
Each option begins with the '\-' character.
(On VAX/VMS or MS-DOS systems you may use either '/' or '\-'.)
For the sake of conformity with an increasingly common convention, options can
also begin with '\-\-'.
The options are described at greater length in the next section.
.PP
\fBftnchek\fP options fall into two categories: switches, which are either
true or false, and settings, which have a numeric or string value.  The
name of a switch is prefixed by 'no' or 'no-' to turn it off:
e.g. \fB\-nopure\fP
would turn off the warnings about impure functions. The 'no' prefix
can also be used with numeric settings, having the effect of turning
off the corresponding warnings.
Settings that control lists of warnings have a special syntax discussed
below.
Only the first 3
characters of an option name (not counting
the '\-')
need be provided.
A colon may be used in place of an equals sign for numeric or string setting
assignments; however, we show only the equals sign form below.
.PP
The switches and settings which \fBftnchek\fP currently recognizes are
listed below.  For each option, the default is the value used if the
option is not explicitly specified, while the turn-on is the value
used if the option is given without assigning it a value.
.TP
.BI \-arguments= list
Control warnings about subprogram type and argument mismatches.
Default = turn-on = all.
.TP
.BI \-array= list
Control warnings in checking array arguments of subprograms.
Default = turn-on = all.
.TP
.BI \-brief
Use shorter format for some error messages.  Default = no.
.TP
.BI \-calltree= list
Produce subprogram call hierarchy in one of 3 formats:
text call-tree, who-calls-who and VCG.
Default = none, turn-on = tree,prune,sort.
.sp 1
If the \fB\-mkhtml\fP option is invoked \fIand\fP tree is the applied
calltree option, a file named \fCCallTree.html\fR, will be produced
depicting the tree in HTML format.
.TP
.B \-check
Perform checking.  Default = yes.
.TP
.BI \-columns= num
Set maximum line length to \fInum\fP columns. (Beyond
this is ignored.)  Turn-on = max = 132.  Default = 72.
.TP
.BI \-common= list
Set degree of strictness in checking COMMON blocks.
Default = turn-on = all.
.TP
.BI \-crossref= list
Print cross-reference list of subprogram calls, label usage, and/or
COMMON block use.  Default = none.
.TP
.B \-declare
Print a list of all identifiers whose datatype is not
explicitly declared.  Default = no.
.TP
.B \-division
Warn wherever division is done (except division by a
constant).  Default = no.
.TP
.BI \-errors= num
Set the maximum number of error messages per cascade.  Default =
turn-on = 3.
.TP
.B \-extern
Warn if external subprograms which are invoked
are never defined.  Default = yes.
.TP
.BI \-f77= list
Control specific warnings about supported extensions to the Fortran 77
Standard.  Default  = none, turn-on = all.
.TP
.BI \-f90= list
Control specific warnings about supported extensions to the Fortran 77
Standard that were not adopted as part of the Fortran 90 Standard.
Default  = none, turn-on = all.
.TP
.BI \-f95= list
Control specific warnings about standard Fortran 77
features that were deleted from the Fortran 95 Standard.  Default  =
none, turn-on = all.
.TP
.B \-help
Print command summary.  Default = no.
.TP
.BI \-identifier\-chars= list
Define non-alphanumeric characters that may be used in identifiers.
Default = turn-on = dollar sign and underscore.
.TP
.BI \-include= path
Define a directory to search for INCLUDE files before searching in the
system-wide directory.
Cumulative.  Default = turn-on = none.
.TP
.BI \-intrinsic= list
Control treatment of nonstandard intrinsic functions.  Default =
all except \fBvms\fP for Unix version, all except \fBunix\fP
for VMS version, all except \fBunix\fP and \fBvms\fP for other
versions.  Turn-on = all.
.TP
.B \-library
Begin library mode: do not warn about subprograms
in file that are defined but never used.  Default = no.
.TP
.B \-list
Print source listing of program.  Default = no.
.TP
.BI \-makedcls= list
Prepare a file of declarations.  The \fIlist\fP specifies options for
the format of this file.  Default = none, turn-on = declarations.
.TP
.BI \-mkhtml= list
Create individual HTML document files from ftnchek analysis and code comments.
Usually you will also want to specify \fB\-call=tree\fP to create the
root HTML file \fCCallTree.html\fR.
Default = none, turn-on = documents.
.TP
.B \-novice
Give output suitable for novice users.  Default = yes.
.TP
.BI \-output= filename
Send output to the given file.
Default and turn-on sends output to the screen. (Default filename
extension is \fI.\|lis\fP).
.TP
.BI \-pointersize= num
Set the size of ``Cray pointer'' variables to \fInum\fP bytes.  Min =
1, max = 16.  Default = turn-on = 4
.TP
.BI \-portability= list
Warn about non-portable usages.  Default = none, turn-on = all.
.TP
.BI \-pretty= list
Give warnings for possibly misleading appearance
of source code.    Default = turn-on = all.
.TP
.BI \-project= list
Create project file (see explanation below).  Default =
no.
.TP
.B \-pure
Assume functions are pure, i.e. have no side effects.
Default = yes.
.TP
.B \-quiet
Produce less verbose output.  Default = no.
.TP
.B \-reference
Print table of subprograms referenced by each subprogram.
Default = no.
.TP
.B \-resources
Print amount of resources used in analyzing the program.
Default = no.
.TP
.B \-sixchar
List any variable names which clash at 6 characters length.
Default = no.
.TP
.B \-sort
Print list of subprograms sorted in prerequisite order.
Default = no.
.TP
.BI \-source= list
Select source formatting options: fixed or free form, DEC Fortran
tab-formatted lines, VMS-style INCLUDE statement,
UNIX-style backslash escape sequences,
and implicit typing of parameters.  Default = none, turn-on = all.
.TP
.BI \-style= list
Produce extra-picky warnings about obsolescent or old-fashioned
programming constructions.  Default = none, turn-on = all.
.TP
.B \-symtab
Print symbol table and label table for each subprogram.  Default = no.
.TP
.BI \-truncation= list
Check for possible loss of accuracy by truncation.
Default = turn-on = all.
.TP
.BI \-usage= list
Control warnings about unused or uninitialized variables, common
blocks, etc.  Default = turn-on = all.
.TP
.B \-vcg
Produce VCG format of call graph.
.TP
.B \-version
Print version number.  Default = no.
.TP
.B \-volatile
Assume COMMON blocks lose definition between
activations.  Default = no. (Obsolete.  Use \fB\-common=volatile\fP instead.) 
.TP
.BI \-wordsize= num
Set the default word size for numeric
quantities to \fInum\fP bytes.  Default = turn-on = 4 bytes.
.TP
.BI \-wrap= num
Set output column at which to wrap long error messages and
warnings to the next line.  If set to 0, turn off wrapping.  Default =
turn-on = 79.
.PP
When more than one option is used, they should be separated by a blank
space, except on systems such as VMS where options begin with slash ( / ).
No blank spaces may be placed around the equals sign ( = ) in a setting.
\fCftnchek "?"\fR
will produce a command summary listing all options and settings.
.PP
For settings that take a list of keywords, namely
\fB\-arguments\fP,
\fB\-array\fP,
\fB\-calltree\fP,
\fB\-common\fP,
\fB\-crossref\fP,
\fB\-f77\fP,
\fB\-f90\fP,
\fB\-f95\fP,
\fB\-intrinsic\fP,
\fB\-makedcls\fP,
\fB\-mkhtml\fP,
\fB\-portability\fP,
\fB\-pretty\fP,
\fB\-project\fP,
\fB\-source\fP,
\fB\-style\fP,
\fB\-truncation\fP,
and
\fB\-usage\fP,
the list consists of keywords
separated by commas or colons.
If the list of keywords is omitted, the
effect is to set the option to its turn-on value (same as ``all'' in most
cases).
Also, if the
list is omitted, the setting name can be prefixed with \fBno\fP or
\fBno\-\fP to
turn off all the options it controls.  For example, \fB\-f77\fP turns
on all warnings about nonstandard constructions, while
\fB\-nof77\fP turns them all off.    Three special keywords are:
.TP
.B help
Print out all the option keywords controlled by the setting,
with a brief explanation of their meanings.
This keyword cannot be given in a list with other keywords.
.TP
.B all
Set all options. This turns on all options controlled by the setting.
.TP
.B none
Clear all options.  This turns off all options controlled by the setting.
.PP
These three special keywords must be given in full.  For all other
keywords, only as many letters of the keyword as are necessary to
identify it unambiguously need be given, or a wildcard pattern
may be used.  Including a keyword in the
list turns the corresponding option on.
For example, \fB\-f77=intrinsic\fP would
turn on only the warnings about use of nonstandard intrinsic functions.
Prefixing a keyword by \fBno\-\fP turns its option off.
For example,
\fB\-pretty=no\-long\-line\fP turns off warnings about lines
exceeding 72 columns in length while leaving all other warnings about
misleading appearance in effect.
If a setting has default \fBnone\fP, you can turn on all options
except one or two by using \fBall\fP first.
For example,
\fB\-f77=all,no\-include\fP enables warnings about all nonstandard
extensions except INCLUDE statements.
If a setting has default \fBall\fP, you can turn off all warnings
except one or two by using \fBnone\fP first.
For example,
\fB\-truncation=none,demotion\fP would turn off all precision related
warnings except about demotions.
Wildcard patterns contain an asterisk to stand for any string of
characters.  If a wildcard pattern is used, all the warnings that
match it are affected.  If \fBno\-\fP is prefixed to the pattern, all
the matching warnings are turned off, otherwise they are all turned
on.  The minimum unambiguous length rule does not apply to wildcard
matching.    For example, use \fB\-usage=no\-*var*\fP to turn off
all warnings relating to variable usage (both local and common).
(Unix users may need to
quote any options containing wildcards in order to prevent the shell
from attempting to expand them.)
Wildcards are recognized only in lists of warning keywords, not in the
top-level options themselves.
.PP
When \fBftnchek\fP starts up, it looks for environment variables
and also for a preferences file.
Any options defined in the environment or in the preferences file are
used as defaults in place of the built-in defaults.  They are
over-ridden by any command line options.
See the section on changing the defaults for details about the
environment options and the preferences file.
.PP
When giving a name of an input file, the extension is optional.  If no
extension is given, \fBftnchek\fP will first look for a project file
with extension \fI.\|prj\fP, and will use that if it exists.  If not, then
\fBftnchek\fP will look for a Fortran source file
with the extension \fI.\|for\fP for VMS systems, \fI.\|f\fP for UNIX systems.
More than one file name can be given to \fBftnchek\fP,
and it will process the
modules in all files as if they were in a single file.
.PP

ftnchek.el
;;; ftnchek.el --- ftnchek support for fortran mode.
;;
;; Author: Judah Milgram <milgram@cgpp.com>

(defvar ftnchek-mode-version "0.9")
(defvar ftnchek-mode-date "12/16/2002")

;; Keywords: fortran syntax semantic
;; Current version at: http://www.glue.umd.edu/~milgram/ftnchekel.html
;;
;; Copyright 1998-2002 Judah Milgram
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc.,  59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.
;;
;; ==================================================================
;;
;;; FTNCHEK: Ftnchek is a fortran 77 syntax and  semantics checker
;;  by Dr. Robert Moniot, <moniot@fordham.edu>. Get it at
;;  http://www.dsm.fordham.edu/~ftnchek/
;;
;;======================================================================
;;   
;;  INSTALLATION:
;;
;;  Install ftnchek.el somewhere in your lisp load path. Maybe add
;;  lines in your ~/.emacs along the lines of:
;;
;;  (setq my-path (concat (getenv "HOME") "/local/share/emacs/site-lisp"
;;  (setq load-path (cons my-path load-path))
;;  (add-hook 'fortran-mode-hook (require 'ftnchek-mode "ftnchek"))
;;
;;  Byte-compile ftnchek.el, if you want.
;;
;; To Do:
;;
;; How do we handle case were comments precede first subroutine in
;;         library file? ftnchek-mode thinks it's an unnamed main.
;; Soup up regexps to tolerate embedded blanks.
;; Documentation! Info file, etc. (for  ftnchek too!)
;; Splash-blurb if ftnchek not found; message where to get
;; Make ftnchek-flags easier for user to customize (one for buffer,
;;         one for subprogram)
;; 
;; ====================================================================
;;
;; Acknowledgements:
;; Bruce Ravel, Jinwei Shen, Richard Stallman, and many others for advice,
;; suggestions and testing. 
;; Also: Michael D. Prange and Dave Love for fortran mode
;; And especially: Bob Moniot for ftnchek!
;;                 
;; ====================================================================
;;  History:
;;  v 0.9 12/16/02  update Bob Moniot contact info
;;                update acks
;;                fix message bug in ftnchek-check-subprogram
;;                clean up some comments
;;                improved ftnchek-error-first-line
;;                many setq's changed to defvar
;;                simplified ftnchek-current-subprogram
;;                consolidated dangling parentheses   :)
;;                miscellaneous cleanup to permit byte-compile w/o warnings
;;  v 0.8 12/10/02 oops, regexp-opt causes problems, switch to regexp-or
;;  v 0.7 12/4/02 Tested with emacs 21
;;                Menu-bar renamed "Ftnchek" and simplified
;;                much internal cleanup and re-write
;;                "next-error" in compile mode works much better now
;;                removed some functions that are now in Fortran-mode
;;                re-did menu with "easy-menu"
;;                pipe ftnchek through sed to make file name look right
;;  v 0.6 6/17/98 placed completion-ignore-case in a let
;;                defvar ftnchek-mode
;;                defun ftnchek-mode
;;  V 0.5 6/14/98 implemented "ftnchek-next-error"
;;                played with ftnchek-flags (array=2) 
;;  V 0.4 6/12/98 added require to "compile"
;;                got "fortran-goto-subprogram" working
;;  V 0.3 6/11/98 first public release

(require 'fortran)
(require 'compile)

(defvar ftnchek-maintainer "<milgram@cgpp.com>")
(defvar ftnchek-flags nil)
(defvar ftnchek-startup-message) ; maybe do this with "let"?

(defvar ftnchek-mode nil
  "Mode variable for ftnchek minor mode")
(make-variable-buffer-local 'ftnchek-mode)

(defcustom ftnchek-buffer-flags
  "-arguments -noextern -declare -library -noarray -portability -usage=no-com-var-uninitialized -include=."
  "Ftnchek options to use when checking an entire buffer")
(defcustom ftnchek-subprogram-flags
  "-arguments -noextern -declare -library -noarray -portability -usage=no-com-var-uninitialized -include=."
  "Ftnchek options to use when checking an individual subprogram")
(defcustom ftnchek-f77-flags
  nil
  "F77 strictness flags that get toggled in pull-down menu")

(defun ftnchek-mode(&optional arg)
  "Ftnchek minor mode."
  (interactive "P")
  (setq ftnchek-mode
	(if (null arg)
	    (not ftnchek-mode)
	  (> (prefix-numeric-value arg)  0))))
;   (if ftnchek-mode  ... etc.

(defun ftnchek-temp-file(s)
  "Generate a temp file with .f suffix"
  (concat
   (make-temp-name
    (expand-file-name s temporary-file-directory))
   ".f"))

(defun ftnchek-delete-lines-forward()
  "Delete all lines starting with current line"
  (save-excursion
    (let ((begin (point))
	  (end (point-max)))
      (delete-region begin end))))

(defun ftnchek-mask-lines-before-here()
  "Replace all lines preceding point with blank lines"
 (save-excursion
   (while (= (forward-line -1) 0)
     (beginning-of-line)
     (let ((beg (point)))
       (end-of-line)
       (delete-region beg (point))))))

(defvar
 ftnchek-error-regexp-alist
 (list
;; line 1 col 2 file foo.f
  (list ".*line \\([0-9]+\\)\\( col \\([0-9]+\\)\\)? file \\([^ ;$|:\n\t]+\\)" 4 1 3)
;; "foo.f", line 14 col 19:
  (list "\"\\([^\"]+\\)\", \\(near \\)?line \\([0-9]+\\)\\( col \\([0-9]+\\)\\)?" 1 3 5)))


(defun ftnchek-region(ftnchek-flags)
  "Run ftnchek on a region using compile()"
  ;; first, last are character positions. Convert to line positions.
  (let ((temp-file (ftnchek-temp-file "ftnchek" ))
	(first (point))
	(last (mark))	)
    (copy-region-as-kill (point-min) (point-max))
    (with-temp-file temp-file
      (yank)
      (goto-char last)
      (ftnchek-delete-lines-forward)
      (goto-char first)
      (ftnchek-mask-lines-before-here)
      )
    (compile-internal (ftnchek-command temp-file ftnchek-flags (buffer-name)) "No more errors" nil nil ftnchek-error-regexp-alist nil nil nil nil )))

(defun ftnchek-command(file-name &optional flags real-name)
  "Form the command to run ftnchek"
  ;; begin and end are line numbers, not char numbers.
  ;; Start by sending file-name to stdout, possibly
  (let ((rval "ftnchek "))
    (if (not (eq flags nil))
	(setq rval (concat rval flags " "))
      )
    (if (not (eq ftnchek-f77-flags nil))
	(setq rval (concat rval ftnchek-f77-flags " "))
      )
    (setq rval (concat rval "-quiet " file-name))
    (if (not (eq real-name nil))
	;; a bit dangerous -
	;; we assume this means file-name is a temp file
	;; maybe not always the case !
      (setq rval (concat rval " | sed 's|" file-name
			 "|" real-name "|g' && rm -f " file-name)))
    rval))

(defun ftnchek-buffer()
  "Run ftnchek on current buffer."
  (interactive)
  (save-excursion
    (mark-whole-buffer)
    (ftnchek-region ftnchek-buffer-flags)
    (message "Checking entire buffer %s" (buffer-name))))

(defun ftnchek-subprogram()
  "Run ftnchek on suprogram the cursor is in. You can run
   fortran-what-subprogram  to find out what subprogram that is."
  (interactive)
  (save-excursion
;; Use this for older versions of fortran-mode:
;;    (mark-fortran-subprogram)
;; As of fortran mode v 21.2 or maybe even earlier:
    (mark-defun)
    (ftnchek-region ftnchek-subprogram-flags))
    (message "Checking %s" (ftnchek-current-subprogram)))


(defun ftnchek-strict-f77()
  "Toggle on strict Fortran 77 compliance checking"
  (interactive)
  (if (equal ftnchek-f77-flags "-f77")
      (setq ftnchek-f77-flags nil)
    (setq ftnchek-f77-flags "-f77")))

; I'm not sure I like these but nobody's complaining.
(define-key fortran-mode-map "\C-x`" 'ftnchek-next-error)
(define-key fortran-mode-map "\M-s" 'ftnchek-subprogram)
(define-key fortran-mode-map "\M-b" 'ftnchek-buffer)
(define-key fortran-mode-map "\M-p" 'ftnchek-previous-subprogram)
(define-key fortran-mode-map "\M-n" 'ftnchek-next-subprogram)
(define-key fortran-mode-map "\M-f" 'ftnchek-first-executable)
(define-key fortran-mode-map "\M-h" 'ftnchek-what-subprogram)



;; Menu
;; Fortran-mode does this, but is it important for us too?
;;(unless (boundp 'ftnchek-mode-menu)  
  (easy-menu-define
   ftnchek-mode-menu fortran-mode-map "Ftnchek menu"
   '("Ftnchek"
     ["Check buffer   " ftnchek-buffer t]
     ["Check subprogram   " ftnchek-subprogram t]
     ["Next error   " ftnchek-next-error t]
     ["Ftnchek version   " ftnchek-version-display t]
     ["Strict F77   " ftnchek-strict-f77 :style toggle
        :selected (equal ftnchek-f77-flags "-f77") ]
     "----"
     ;;; These items really belong in the fortran mode menu:
     ["What subprogram?   " ftnchek-what-subprogram t]
     ["First executable   " ftnchek-first-executable t]
     ["Prev subprogram   " ftnchek-previous-subprogram t]
     ["Next subprogram   " ftnchek-next-subprogram t]
     ))
 ;; )


; Startup message. Possibly useless.
(setq ftnchek-startup-message
      (concat "ftnchek.el "
	      " Version "
	      ftnchek-mode-version
	      " "
	      ftnchek-mode-date
	      " bugs to "
	      ftnchek-maintainer))
(message ftnchek-startup-message)
(sleep-for 0.5)

(defun ftnchek-version-display()
"Print the ftnchek version and patch level."
(interactive)
(message (concat (ftnchek-version) "; ftnchek.el v. " ftnchek-mode-version)))

;; This should probably be done with a pipe and sed.
(defun ftnchek-version()
  "Return ftnchek version as a string."
  (let (first last outbuf)
    (setq outbuf (get-buffer-create "*Ftnchek*"))
    (set-buffer outbuf)
    (goto-char (point-min))
    (setq first (point))
    (goto-char (point-max))
    (setq last (point))
    (if (> last first) (kill-region first last))
    (call-process "ftnchek" nil outbuf nil "-help")
    (set-buffer outbuf)
    (goto-char (point-min))
    (if (null (search-forward "FTNCHEK")) nil
      (beginning-of-line)
      (setq first (point))
      (end-of-line)
      (setq last (point))
      (buffer-substring first last))))

(defun ftnchek-error-first-line()
  "set first line of multiline ftnchek error message to top of window"
  (let (( here (point)))
    (beginning-of-line)
    (if (not (looking-at "^.*\\(Warning\\|Error\\)"))
	(re-search-backward "^.*\\(Warning\\|Error\\)" nil t)
      (forward-line -1)
      (if (not (looking-at "^ *\\^"))
	  (goto-char here)
	(forward-line -1)
	(if (not (looking-at "^ *[0-9]+"))
	    (forward-line 2))
	)))
  (recenter 0))

(defun ftnchek-next-error()
"ftnchek mode wrapper for next-error"
(interactive)
(next-error)
(other-window 1)
(ftnchek-error-first-line) 
(other-window -1))


;; I hope this is a good idea
(setq compilation-error-regexp-alist
      (append ftnchek-error-regexp-alist
	    compilation-error-regexp-alist))



(provide 'ftnchek-mode)


;;; ***********************************************************************
;
;     Extra navigation stuff - maybe this functionality will be added to
;     Fortran mode, in which case we can drop it from here.
;
;     Note ftnchek-mode's idea of where a program unit begins or ends may
;     not agree with fortran-mode.

;;    some useful regexps:

(defun identity(x) x)
(defun regexp-or(s &optional parens)
  "OR together a bunch of regexp's. Optional argument if t adds outer parens"
  (let ((rval (mapconcat 'identity s "\\|")))
    (if (eq parens nil)
	rval
      (concat "\\(" rval "\\)"))))

;; What about embedded spaces?

(defvar ftnchek-first-six-regexp "^[0-9 ][0-9 ][0-9 ][0-9 ][0-9 ] +")
(defvar ftnchek-blank-line-regexp "^[ \t]*$")      
(defvar ftnchek-continuation-line-regexp "^[0-9 ][0-9 ][0-9 ][0-9 ][0-9 ][^ ] *")
(defvar ftnchek-comment-regexp "^[Cc]" )
(defvar ftnchek-symbolic-name-regexp "\\([a-zA-Z][a-zA-Z0-9]*\\)")

(defvar ftnchek-type-regexp-list (list
				"integer"
				"real"
				"double *precision"
				"complex"
				"double *complex" ; not standard
				"logical"
				"\\(character\\( *\\* *[0-9]+\\)?\\)" ))

(defvar ftnchek-type-regexp (regexp-or ftnchek-type-regexp-list t))

(defvar ftnchek-subprogram-end-regexp (concat ftnchek-first-six-regexp "end *$"))

(defvar ftnchek-program-unit-begin-regexp
      (concat ftnchek-first-six-regexp
	      (regexp-or (list
			  (concat ftnchek-type-regexp "? *function")
			  "subroutine"
			  "program"
			  "block *data") t)
	      " *" ftnchek-symbolic-name-regexp "?" ))

;; See F77 standard, section 7. Note this regexp can't pick up
;; statement functions, which F77 also classes as non-executable.

; The items commented out are covered in ftnchek-type-regexp
(defvar ftnchek-non-executable-keyword-regexp-list
      (list
       "block *data"
       "character"
       "common"
       "complex"
       "data"
       "dimension"
       "function"
       "double *complex" ; not strict f77
       "double *precision"
       "entry"
       "equivalence"
       "external"
       "format"
       "implicit"
       "include"   ; not strict f77
       "integer"
       "intrinsic"
       "logical"
       "parameter"
       "program"
       "real"
       "save"
       "subroutine"))

(defvar ftnchek-non-executable-statement-regexp
      (concat ftnchek-first-six-regexp
	      (regexp-or ftnchek-non-executable-keyword-regexp-list t)))


(defun  ftnchek-program-unit-title()
  "Return descriptive string for program unit, or nil"
  (interactive)
  (save-excursion
    (beginning-of-line)
    (if (not (looking-at ftnchek-program-unit-begin-regexp))
	nil
      ;; Guess how I got those match-field numbers.
      (let (( title  (match-string 1) )
	    ( name   (match-string 5) ))
	(if (not (eq name nil))
	    (concat title " " name)
	  name)))))


(defun ftnchek-end-of-subprogram()
  "Move point to first character of end statement (or EOF)."
  (let (( here (point) ))
    (if (re-search-forward ftnchek-subprogram-end-regexp nil 1 )
	(beginning-of-line)
      (if (re-search-backward ftnchek-subprogram-end-regexp nil t)
	  (beginning-of-line)
	(message "No end statement found beyond this point.")))
      (goto-char here)))

(defun ftnchek-find-program-unit-statement( N )
  "Move point either forwards or backwards to program unit start statement,
and return the title, or nil. N is either 1 (forward) or -1 (backward)"
  (interactive "p")
  (beginning-of-line)
  (let ((name nil))
    (while (and (not (setq name (ftnchek-program-unit-title)))
		(eq (forward-line N) 0)))
    (if (and (eq name nil)
	     (eq N -1))
	(setq name "unnamed main program")
      )
    name ))

(defun ftnchek-beginning-of-subprogram()
  "Move point to start of a program unit. Could be beginning of file.
   Returns title of program unit."
  (interactive)
  (beginning-of-line)
  ;; interstitial comments belong to following subprogram
  (ftnchek-end-of-subprogram)
  (ftnchek-find-program-unit-statement -1)
  )

(defun ftnchek-current-subprogram()
  "Return name of current subprogram without actually moving point"
  (save-excursion
    (ftnchek-beginning-of-subprogram)))
 
(defun ftnchek-what-subprogram()
  "Display the title of current Fortran subprogram"
  (interactive)
  (message (ftnchek-current-subprogram)))

(defun ftnchek-next-subprogram()
  "Move point to next subprogram"
  (interactive)
  (let (( here (point)) )
    (forward-line 1)
    (if (ftnchek-find-program-unit-statement 1)
	(message (ftnchek-current-subprogram))
      (message "Don't seem to be any more" )
      (goto-char here))))

(defun ftnchek-previous-subprogram()
  "Move point to previous subprogram"
  (interactive)
  ;; moving backwards, we always get a program unit name even if "unnamed main"
  (if (not (re-search-backward ftnchek-subprogram-end-regexp nil t))
	(message "Already seem to be in first one")
      (ftnchek-beginning-of-subprogram)
      (message (ftnchek-current-subprogram))))
  


(defun ftnchek-nonexecutable-statement()
  "t if current line is nonexecutable"
;; let's give it an optional arg so we can look at strings
  (or (looking-at ftnchek-non-executable-statement-regexp)
      (looking-at ftnchek-comment-regexp)
      (looking-at ftnchek-continuation-line-regexp)
      (looking-at ftnchek-blank-line-regexp)))

(defun ftnchek-next-executable-statement()
  "Skip to next executable statement"
  (while (ftnchek-nonexecutable-statement) (forward-line)))

(defun ftnchek-first-executable()
  "Move cursor to first executable statement in current subprogram"
  (interactive)
  (ftnchek-beginning-of-subprogram)
  (ftnchek-next-executable-statement)
  (message "First executable statement in %s" (ftnchek-current-subprogram)))
